In [ ]:
!pip install -q pyomo
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
#Creating a model
model1 = ConcreteModel()

#No. of decision variables
N = 20
M = 3

#Objective function coefficient using numpy array
obj_coef_c = np.array([45, 48, 55, 65, 54, 37.5, 35, 52.5, 47, 28, 82, 102, 150, 105, 95, 84, 88, 125, 70, 64])

#Constraints coeffiecient using numpy array
constr_coef_A = 0.01*np.array([20, 25, 10, 30, 12.5, 35, 47.5, 10, 20, 50, 20, 25, 10, 30, 12.5, 35, 47.5, 10, 20, 50, 75, 65, 70, 60, 85, 52.5, 42.5, 40, 35, 40, 75, 65, 70, 60, 85, 52.5, 42.5, 40, 35, 40, 5, 10, 20, 10, 2.5, 12.5, 10, 50, 45, 10, 5, 10, 20, 10, 2.5, 12.5, 10, 50, 45, 10,
]).reshape(3, 20)
constr_coef_B = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [ ]:
#Declaring RHS of the constraints as variables
constr1_rhs_b = (425/100)*np.array([27.5, 57.5, 15])

#Declaring lower and upper bounds of the constraints
l = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
u = np.array([15, 16, 9, 17.5, 12, 18.5, 2.5, 0.5, 9.5, 8, +np.inf, +np.inf, +np.inf, +np.inf, +np.inf, +np.inf, +np.inf, +np.inf, +np.inf, +np.inf, ])

#Set of row and column indices
row_indices = np.arange(M)
col_indices = np.arange(N)
print(col_indices)
print(col_indices[1])

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
1


In [ ]:
#Declaring the decision variables in model
model1.x = Var(col_indices)

In [ ]:
#Creating a constraint list to hold multiple constraints
model1.constraints = ConstraintList()

#Adding model constraints one by one to the list
#first add the constraint p1x1 + p2x2 +.......+ pNxN == b1
for i in row_indices:
 model1.constraints.add(sum(constr_coef_A[i][j]*model1.x[j] for j in col_indices) == constr1_rhs_b[i])
#second add the constraints q1x1 + qzx2 +......+ qNxN == b2
 model1.constraints.add(sum(constr_coef_B[j]*model1.x[j] for j in col_indices) == 425)

In [ ]:
#add the upper and lower bounds for the variables
for j in col_indices:
  model1.x[j].setlb(l[j])
  model1.x[j].setub(u[j])

In [ ]:
#add the model objective
model1.objective = Objective(expr = sum(obj_coef_c[j]*model1.x[j] for j in col_indices), sense = minimize)

In [ ]:
#print the whole model
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   20 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}

1 Var Declarations
    x : Size=20, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  15.0 : False :  True :  Reals
          1 :     0 :  None :  16.0 : False :  True :  Reals
          2 :     0 :  None :   9.0 : False :  True :  Reals
          3 :     0 :  None :  17.5 : False :  True :  Reals
          4 :     0 :  None :  12.0 : False :  True :  Reals
          5 :     0 :  None :  18.5 : False :  True :  Reals
          6 :     0 :  None :   2.5 : False :  True :  Reals
          7 :     0 :  None :   0.5 : False :  True :  Reals
         

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 28747.1875
  Upper bound: 28747.1875
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 21
  Number of nonzeros: 121
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.013132572174072266
# ----------------------------------------------------------
#   Solution Information
# -----------------------

In [ ]:
#display solution
print('\nObjective = ', model1.objective())
print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model1.x[i].value)
print('\nConstraints')
model1.constraints.display()


Objective =  28747.1875

Decision Variables
x[ 0 ] =  15.0
x[ 1 ] =  16.0
x[ 2 ] =  9.0
x[ 3 ] =  17.5
x[ 4 ] =  12.0
x[ 5 ] =  18.5
x[ 6 ] =  2.5
x[ 7 ] =  0.5
x[ 8 ] =  9.5
x[ 9 ] =  8.0
x[ 10 ] =  158.078125
x[ 11 ] =  0.0
x[ 12 ] =  0.0
x[ 13 ] =  0.0
x[ 14 ] =  0.0
x[ 15 ] =  0.0
x[ 16 ] =  0.0
x[ 17 ] =  0.0
x[ 18 ] =  74.046875
x[ 19 ] =  84.375

Constraints
constraints : Size=6
    Key : Lower   : Body    : Upper
      1 : 116.875 : 116.875 : 116.875
      2 :   425.0 :   425.0 :   425.0
      3 : 244.375 : 244.375 : 244.375
      4 :   425.0 :   425.0 :   425.0
      5 :   63.75 :   63.75 :   63.75
      6 :   425.0 :   425.0 :   425.0


**Report of exercise 3 part 1:**
For formulating the LP for this problem we'll be using this generalised method:

      min (c^T)x,

      s.t Ax ≤ b

       l ≤ x ≤ u


*  Here we have taken 20 decision variables {x1, x2,........, x20} in the variables numbered from x1 to x10 are the variables denoting in-house stock of alloys.

*   While the variables numbered from x11 to x20 are the variables denoting the purchasing alloys.

*   The first three rows of the table given in the question itself are the constraint coefficients.

*   The in-house stock(4th row) would be the upper bound for all the in-house alloys respectively.

*   The last two rows would be contributing as objective function coefficients.

*   While the upper bounds for all the purchased alloys would be infinity as they are unlimited in quantity.

*   The lower bounds for all the 20 variables would be 0.0 .

*   The matrix **"b"** would be the required composition of chromium, aluminium and silver.



 **So, following these points, the formulation would look like as follows:**

**Objective function:**

      45.0*x[0] + 48.0*x[1] + 55.0*x[2] + 65.0*x[3] + 54.0*x[4] + 37.5*x[5] +35.0*x[6] + 52.5*x[7] + 47.0*x[8] + 28.0*x[9] + 82.0*x[10] + 102.0*x[11] + 150.0*x[12] + 105.0*x[13] + 95.0*x[14] + 84.0*x[15] + 88.0*x[16] + 125.0*x[17] + 70.0*x[18] + 64.0*x[19]

**Constraints:**

*   0.2*x[0] + 0.25*x[1] + 0.1*x[2] + 0.3*x[3] + 0.125*x[4] + 0.35000000000000003*x[5] + 0.47500000000000003*x[6] + 0.1*x[7] + 0.2*x[8] + 0.5*x[9] + 0.2*x[10] + 0.25*x[11] + 0.1*x[12] + 0.3*x[13] + 0.125*x[14] + 0.35000000000000003*x[15] + 0.47500000000000003*x[16] + 0.1*x[17] + 0.2*x[18] + 0.5*x[19] == 116.875


* x[0] + x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7] + x[8] + x[9] + x[10] + x[11] + x[12] + x[13] + x[14] + x[15] + x[16] + x[17] + x[18] + x[19] == 425

*  0.75*x[0] + 0.65*x[1] + 0.7000000000000001*x[2] + 0.6*x[3] + 0.85*x[4] + 0.525*x[5] + 0.425*x[6] + 0.4*x[7] + 0.35000000000000003*x[8] + 0.4*x[9] + 0.75*x[10] + 0.65*x[11] + 0.7000000000000001*x[12] + 0.6*x[13] + 0.85*x[14] + 0.525*x[15] + 0.425*x[16] + 0.4*x[17] + 0.35000000000000003*x[18] + 0.4*x[19] == 244.375

*    0.05*x[0] + 0.1*x[1] + 0.2*x[2] + 0.1*x[3] + 0.025*x[4] + 0.125*x[5] + 0.1*x[6] + 0.5*x[7] + 0.45*x[8] + 0.1*x[9] + 0.05*x[10] + 0.1*x[11] + 0.2*x[12] + 0.1*x[13] + 0.025*x[14] + 0.125*x[15] + 0.1*x[16] + 0.5*x[17] + 0.45*x[18] + 0.1*x[19] == 63.75

*   Xi >= 0 for all i in [1, 20]

*   x1 <= 15, x2 <= 16, x3 <= 9, x4 <= 17.5, x5 <= 12, x6 <= 18.5, x7 <= 2.5, x8 <= 0.5, x9 <= 9.5, x10 <= 8

*   x11, x12, ........, x20 <= infinity





















**Report on exercise 3 part 3:**


*   **The value of the decision variables at the optimal point are:**

     x1 = 15.0

     x2 =  16.0

     x3 =  9.0

     x4 =  17.5

     x5 =  12.0

     x6 =  18.5

     x7 =  2.5

     x8 =  0.5

     x9 =  9.5

     x10 =  8.0

     x11 =  158.078125

     x12 =  0.0

     x13 =  0.0

     x14 =  0.0

     x15 =  0.0

     x16 =  0.0

     x17 =  0.0

     x18 =  0.0

     x19 =  74.046875

     x20 =  84.375

*   **The value of the objective fuction at the optimal point is:**

       28747.1875




**Report of exercise 3 part 4:**

*   Among the alloys that need to be bought from the market and are ranked high are:
   A-1, A-9 and A-10 of which A-1 is ranked highest and then A-10 and then A-19.


*   Among all the alloys that need to be bought from from the market A-2, A-3, A-4, A-5, A-6, A-7, A-8 all are ranked low because they actually don't need to be bought.






**Report of exercise 3 part 5:**


*   All the stocked alloys are completely utilized since the cost of in-house alloys and purchasing alloys differ much, so the company must have been decided to finish the in-house alloys first.
  



**Report of exercise 3 part 6:**


*  Alloys A-2 to A-8 need not to be acquired from market because the in-house stocked alloys were enough for the company.

